### PR TITLE
feat: implement dttm column configuration through db extra config

### PR DESCRIPTION
**Pull request by bkyryliuk**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Provides an ability for the superset admin to configure the date time columns per name so they will be automatically created on the table creation.
It helps a lot for the companies that have conventions & automatic day columns.

Reimplements https://github.com/apache/incubator-superset/pull/9353 to be more in line with the https://github.com/apache/incubator-superset/issues/6360

### TEST PLAN
* unit tests
* local testing

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@john-bodley @villebro

### CODE REVIEW

File: docs/installation.rst
Comment (john-bodley) [inline docs/installation.rst:1382]: I wonder if this should be named `main_dttm_col` to be consistent with the datasource level field.

File: superset/connectors/sqla/models.py
Comment (john-bodley) [inline superset/connectors/sqla/models.py:151]: Why not just `return ( ...`. This results in one less variable, etc. and thus make the code easier to grok.

File: docs/installation.rst
Comment (john-bodley) [reply docs/installation.rst:1382]: I sense these default settings (including the `python_date_format_by_column_name` should be documented outside of the SIP-15 section (this section can then reference said section). This would help to provide context regarding default database specific fields which may be inherited either by new Superset tables going forward or used as a fallback (in the case of SIP-15) if a value isn't defined.

Additionally I was wondering if having all these nested under a `defaults` section ensures more consistency (and thus no need for the `default_` prefix), 

```json
{
    "defaults": {
        "dttm_column_names": ["ds", "hour_ts"],
        "main_dttm_col": "ds",
        "expression_by_column_name": {
            "hour_ts": "CAST(hour_ts as INTEGER)"
        },
        "python_date_format_by_column_name": {
            "ds": "%Y-%m-%d",
            "hour_ts": "epoch_s"
        }	       
    }
}
```

would be better documented in their own section, i.e., outside of SIP-15. It could then be clearer to the reader that these database level overrides

File: docs/installation.rst
Comment (john-bodley) [inline docs/installation.rst:1389]: Nit. No trailing comma in JSON.

File: docs/installation.rst
Comment (john-bodley) [inline docs/installation.rst:1386]: Nit. No trailing comma in JSON.

**Review by john-bodley — COMMENTED** [review]
@bkyryliuk overall this looks good. I just had a couple of small comments regarding documentation and naming conventions.

File: docs/installation.rst
Comment (john-bodley) [reply docs/installation.rst:1389]: I think this example is a tad misleading as you woudn't have the same column name using both the Python datetime format and expression as these are mutually exclusive.

**Comment by john-bodley** [comment]
@bkyryliuk one other question I had, and maybe the community should weigh in here, is whether i) defaults should be persisted on tables etc. or ii) whether undefined values should fallback to the defaults.

File: tests/core_tests.py
Comment (willbarrett) [inline tests/core_tests.py:1174]: Can we break down this test?

File: tests/core_tests.py
Comment (willbarrett) [inline tests/core_tests.py:1201]: Comments like this in a test file usually indicate to me that the test should be broken up, and that the comments are a good place to do the breaking.

File: superset/connectors/sqla/models.py
Comment (etr2460) [inline superset/connectors/sqla/models.py:156]: nit: typo `db_default_expreassion`

**Comment by bkyryliuk** [comment]
@willbarrett and @john-bodley addressed the comments, please let me know if you have other suggestions

File: docs/installation.rst
Comment (willbarrett) [inline docs/installation.rst:1348]: Unnecessary `-`

File: superset/connectors/sqla/models.py
Comment (willbarrett) [inline superset/connectors/sqla/models.py:152]: Is this TODO going to be handled in the near future? If not, I'd recommend removing it.

Comment (bkyryliuk) [reply superset/connectors/sqla/models.py:152]: yes in near future if there are no objections

**Comment by bkyryliuk** [comment]
@willbarrett addressed the comments, thanks

**Comment by bkyryliuk** [comment]
@willbarrett , @john-bodley - is there anything else I can do to get this merged?

**Comment by willbarrett** [comment]
No further comments from me. I recommend pinging a couple more contributors to review.

**Comment by mistercrunch** [comment]
Generally LGTM

I think personally I prefer dynamic configuration hooks, where you'd define a function that receives a live table object upon insert and/or modify, and mutate it with whatever business/semantical logic make sense to your org, keeping all the logic on your side.

Intricate static configuration like the one defined here is hard to discover, often requires to read code to understand what it does / can do, and often has shortcomings / requires code change to the core package to get things done.

**Comment by bkyryliuk** [comment]
Thanks @mistercrunch , I think as an alternative, we could benefit from making them high level db configuration in the ui. It would make things more transparent. However this would be probably out of scope of this PR.

@willbarrett is PR in a good state to merge?

**Comment by willbarrett** [comment]
@bkyryliuk I have no further blockers, but I:

1. Don't understand this part of the system well enough to give you a merge approval
2. Am not a commiter, so my approval carries little weight.

Again, please reach out to additional committers to see if this is viable. It sounds like @mistercrunch would prefer an alternate solution.

**Comment by bkyryliuk** [comment]
> @bkyryliuk I have no further blockers, but I:
> 
> 1. Don't understand this part of the system well enough to give you a merge approval
> 2. Am not a commiter, so my approval carries little weight.
> 
> Again, please reach out to additional committers to see if this is viable. It sounds like @mistercrunch would prefer an alternate solution.

Thanks for the context. Will do.

**Comment by mistercrunch** [comment]
Here's an example of what I meant:
https://github.com/apache/incubator-superset/compare/master...mistercrunch:example_config?expand=1

Using this, you can implement whatever logic to change table configuration based on naming conventions and such.

Would that work for you?

**Comment by bkyryliuk** [comment]
> https://github.com/apache/incubator-superset/compare/master...mistercrunch:example_config?expand=1

I like it, will redo the pr

**Comment by bkyryliuk** [comment]
@mistercrunch  - updated the PR to follow the mutator pattern

**Comment by bkyryliuk** [comment]
test failure doesn't seem to be releated.

**Comment by villebro** [comment]
@bkyryliuk I restarted CI, I had a similar problem yesterday.

**Review by mistercrunch — APPROVED** [review]
LGTM, in the future one flag that could be useful is `is_new` or something like that, for the callback to know whether the datasource is getting created or updated

**Comment by john-bodley** [comment]
@bkyryliuk do you think there is merit in addressing [this](https://github.com/apache/incubator-superset/pull/9444#discussion_r402066069) comment in a follow up PR? I'm a little concerned that there's two different mechanism for defining the Python date-format within the extra blob.

**Pull request merged**
